tests/run_tests.py

```diff
@@ -16,6 +16,82 @@
 GIDIR = 'src/lib'
 
 
+def _get_tests_from_suite(suite, tests):
+    """ Extract tests from the test suite """
+    # 'tests' we get from 'unittest.defaultTestLoader.discover' are "wrapped"
+    # in multiple 'unittest.suite.TestSuite' classes/lists so we need to "unpack"
+    # the indivudual test cases
+    for test in suite:
+        if isinstance(test, unittest.suite.TestSuite):
+            _get_tests_from_suite(test, tests)
+
+        if isinstance(test, unittest.TestCase):
+            tests.append(test)
+
+    return tests
+
+
+def _get_test_tags(test):
+    """ Get test tags for single test case """
+
+    tags = []
+
+    # test failed to load, usually some ImportError or something really broken
+    # in the test file, just return empty list and let it fail
+    # with python2 the loader will raise an exception directly without returning
+    # a "fake" FailedTest test case
+    if six.PY3 and isinstance(test, unittest.loader._FailedTest):
+        return tags
+
+    test_fn = getattr(test, test._testMethodName)
+
+    # it is possible to either tag a test funcion or the class so we need to
+    # check both for the tag
+    if getattr(test_fn, "slow", False) or getattr(test_fn.__self__, "slow", False):
+        tags.append(TestTags.SLOW)
+    if getattr(test_fn, "unstable", False) or getattr(test_fn.__self__, "unstable", False):
+        tags.append(TestTags.UNSTABLE)
+    if getattr(test_fn, "unsafe", False) or getattr(test_fn.__self__, "unsafe", False):
+        tags.append(TestTags.UNSAFE)
+    if getattr(test_fn, "core", False) or getattr(test_fn.__self__, "core", False):
+        tags.append(TestTags.CORE)
+    if getattr(test_fn, "nostorage", False) or getattr(test_fn.__self__, "nostorage", False):
+        tags.append(TestTags.NOSTORAGE)
+    if getattr(test_fn, "extradeps", False) or getattr(test_fn.__self__, "extradeps", False):
+        tags.append(TestTags.EXTRADEPS)
+    if getattr(test_fn, "regression", False) or getattr(test_fn.__self__, "regression", False):
+        tags.append(TestTags.REGRESSION)
+
+    return tags
+
+
+def _print_skip_message(test, skip_tag):
+
+    # test.id() looks like 'crypto_test.CryptoTestResize.test_luks2_resize'
+    # and we want to print 'test_luks2_resize (crypto_test.CryptoTestResize)'
+    test_desc = test.id().split(".")
+    test_name = test_desc[-1]
+    test_module = ".".join(test_desc[:-1])
+
+    if skip_tag == TestTags.SLOW:
+        reason = "skipping slow tests"
+    elif skip_tag == TestTags.UNSTABLE:
+        reason = "skipping unstable tests"
+    elif skip_tag == TestTags.UNSAFE:
+        reason = "skipping test that modifies system configuration"
+    elif skip_tag == TestTags.EXTRADEPS:
+        reason = "skipping test that requires special configuration"
+    elif skip_tag == TestTags.CORE:
+        reason = "skipping non-core test"
+    else:
+        reason = ""  # just to be sure there is some default value
```

Comment:
Something like `"unknown reason"` would be better here.

Reply:
Fixed.